File: src/Python/1-100/81.SearchInRotatedSortedArrayII.py

class Solution:
    def search(self, nums: list, target: int) -> bool:
        left = 0
        right = len(nums) - 1
        if right == -1:
            return False
        k = 0 
        for i in range(1, right + 1):
            if nums[i] < nums[i - 1]:
                k = i
        while left <= right:
            beforemiddle = (left + right) // 2
            middle = beforemiddle + k
            if middle >= len(nums):
                middle = middle - len(nums)
            if target == nums[middle]:
                return True
            elif target < nums[middle]:
                right = beforemiddle - 1
            else:
                left = beforemiddle + 1
        return False


if __name__ == "__main__":
    solution = Solution()
    print(solution.search([2,5,6,0,0,1,2], 0))
    print(solution.search([2,5,6,0,0,1,2], 3))
    print(solution.search([3,1], 1))
    print(solution.search([1,3], 1))
   

File: src/Java/1-100/81.SearchInRotatedSortedArrayII.java

import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Queue;
import java.util.HashSet;
import java.util.LinkedList;

class Solution {
    public boolean search(int[] nums, int target) {
        int left = 0;
        int right = nums.length - 1;
        if (right == -1)
            return false;
        int k = 0;
        for (int i = 1; i < right + 1; i++){
            if (nums[i] < nums[i - 1])
                k = i;
        }
        while (left <= right){
            int beforemiddle = (left + right) / 2;
            int middle = beforemiddle + k;
            if (middle >= nums.length)
                middle = middle - nums.length;
            if (target == nums[middle])
                return true;
            else if (target < nums[middle])
                right = beforemiddle - 1;
            else
                left = beforemiddle + 1;
        }
        return false;
    }
}
